File: react-java/backend/src/main/java/com/clt/hrm/core/administration/program/dtos/ProgramDto.java
package com.clt.hrm.core.administration.program.dtos;

import java.util.List;

import com.clt.hrm.infra.common.dtos.BaseDto;

import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.NoArgsConstructor;

@Data
@NoArgsConstructor
@EqualsAndHashCode(callSuper = true)
public class ProgramDto extends BaseDto {
	private static final long serialVersionUID = -6288035537005720485L;
	private String pgmId;
    private String pgmCd;
    private String pgmNm;
    private String pgmTpCd;
    private String prntPgmId;
    private String pgmRmk;
    private int dspOrder;
    
    // Tree Structure
    private int level;
    private String treeKey;
    private String treePath;
    
    // Permission List
    List<PermissionDto> permList;
}

File: nuxtjs-nestjs/backend/src/apps/administration/programs/dto/program.dto.ts
import {
  IsArray,
  IsIn,
  IsInt,
  IsOptional,
  IsString,
  ValidateNested,
} from 'class-validator';
import { Type } from 'class-transformer';
import { ApiPropertyOptional } from '@nestjs/swagger';
import { PermissionDto } from './permission.dto';
import { BaseDto } from '@infra/common/dto';

export class ProgramDto extends BaseDto {
  @IsOptional()
  @IsString()
  pgmId?: string;

  @IsOptional()
  @IsString()
  pgmCd?: string;

  @IsOptional()
  @IsString()
  pgmNm?: string;

  @IsOptional()
  @IsString()
  @IsIn(['MENU', 'UI'])
  pgmTpCd?: string;

  @IsOptional()
  @IsString()
  prntPgmId?: string;

  @IsOptional()
  @IsInt()
  @Type(() => Number)
  dspOrder?: number;

  @IsOptional()
  @IsString()
  pgmRmk?: string;

  @IsOptional()
  @IsString()
  pgmPath?: string;

  // Computed tree fields (read-only, populated from CTE)
  @IsOptional()
  @ApiPropertyOptional()
  level?: number;

  @IsOptional()
  @ApiPropertyOptional()
  treeKey?: string;

  @IsOptional()
  @ApiPropertyOptional()
  treePath?: string;

  @IsOptional()
  @IsArray()
  @ValidateNested({ each: true })
  @Type(() => PermissionDto)
  permList?: PermissionDto[];
}
